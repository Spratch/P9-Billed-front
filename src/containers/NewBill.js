import { ROUTES_PATH } from '../constants/routes.js'
import Logout from "./Logout.js"

export default class NewBill {
  constructor({ document, onNavigate, store, localStorage }) {
    this.document = document
    this.onNavigate = onNavigate
    this.store = store
    const formNewBill = this.document.querySelector(`form[data-testid="form-new-bill"]`)
    formNewBill.addEventListener("submit", this.handleSubmit)
    const file = this.document.querySelector(`input[data-testid="file"]`)
    file.addEventListener("change", this.handleChangeFile)
    this.fileUrl = null
    this.fileName = null
    this.billId = null
    new Logout({ document, localStorage, onNavigate })
  }
  handleChangeFile = e => {
    e.preventDefault()
    const file = this.document.querySelector(`input[data-testid="file"]`).files[0]
    const filePath = e.target.value.split(/\\/g)
    const fileName = filePath[filePath.length-1]
    const fileNames = fileName.split('.')
    const fileExtension = fileNames[fileNames.length - 1]
    const allowedFiles = ['jpg', 'png', 'jpeg']
    const submitButton = this.document.querySelector(`#btn-send-bill`)
    submitButton.setAttribute("data-testid", "submit-button")

    if (allowedFiles.includes(fileExtension)) {
      const errorMessage = this.document.querySelector(`div[data-testid="file-error"]`)
      if (errorMessage) {
        errorMessage.remove()
      }
      const formData = new FormData()
      const email = JSON.parse(localStorage.getItem("user")).email
      formData.append('file', file)
      formData.append('email', email)
  
      this.store
        .bills()
        .create({
          data: formData,
          headers: {
            noContentType: true
          }
        })
        .then(({fileUrl, key}) => {
          console.log(fileUrl)
          this.billId = key
          this.fileUrl = fileUrl
          this.fileName = fileName
        }).catch(error => console.error(error))
        submitButton.disabled = false
    } else {
      const fileInput = this.document.querySelector(`input[data-testid="file"]`).parentNode
      const errorMessage = this.document.createElement("div")
      errorMessage.setAttribute("data-testid", "file-error")
      errorMessage.textContent = "Uploadez une image en PNG, JPG ou JPEG"
      fileInput.append(errorMessage)
      submitButton.disabled = true;
    }
  }
  handleSubmit = e => {
    e.preventDefault()
    console.log('e.target.querySelector(`input[data-testid="datepicker"]`).value', e.target.querySelector(`input[data-testid="datepicker"]`).value)
    const email = JSON.parse(localStorage.getItem("user")).email
    const allowedFiles = ['jpg', 'png', 'jpeg']
    const fileNames = this.fileName.split('.')
    // const fileNames = ["test","jpg"]
    const fileExtension = fileNames[fileNames.length - 1]
    if (allowedFiles.includes(fileExtension)) {
      const bill = {
        email,
        type: e.target.querySelector(`select[data-testid="expense-type"]`).value,
        name:  e.target.querySelector(`input[data-testid="expense-name"]`).value,
        amount: parseInt(e.target.querySelector(`input[data-testid="amount"]`).value),
        date:  e.target.querySelector(`input[data-testid="datepicker"]`).value,
        vat: e.target.querySelector(`input[data-testid="vat"]`).value,
        pct: parseInt(e.target.querySelector(`input[data-testid="pct"]`).value) || 20,
        commentary: e.target.querySelector(`textarea[data-testid="commentary"]`).value,
        fileUrl: this.fileUrl,
        fileName: this.fileName,
        status: 'pending'
      }
      this.updateBill(bill)
      this.onNavigate(ROUTES_PATH['Bills'])
    }
  }

  // not need to cover this function by tests
  updateBill = (bill) => {
    if (this.store) {
      this.store
      .bills()
      .update({data: JSON.stringify(bill), selector: this.billId})
      .then(() => {
        this.onNavigate(ROUTES_PATH['Bills'])
      })
      .catch(error => console.error(error))
    }
  }
}