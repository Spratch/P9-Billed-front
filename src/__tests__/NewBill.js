/**
 * @jest-environment jsdom
 */

import { fireEvent, getByTestId, screen } from "@testing-library/dom"
import NewBillUI from "../views/NewBillUI.js"
import NewBill from "../containers/NewBill.js"
import user from '@testing-library/user-event';
import mockStore from "../__mocks__/store"
import {localStorageMock} from "../__mocks__/localStorage.js";
import router from "../app/Router.js";
import { ROUTES_PATH} from "../constants/routes.js";
import '@testing-library/jest-dom'


describe("Given I am connected as an employee", () => {
  describe("When I am on NewBill Page", () => {
    test("Then it should renders new bill form page", () => {
      const html = NewBillUI()
      document.body.innerHTML = html

      expect(getByTestId(document.body, 'form-new-bill')).toBeDefined()
    })

    describe("When I try to upload a file", () => {
      beforeEach(() => {
        Object.defineProperty(window, 'localStorage', { value: localStorageMock })

        window.localStorage.setItem('user', JSON.stringify({
          type: 'Employee'
        }))
        const root = document.createElement("div")
        root.setAttribute("id", "root")
        document.body.append(root)
        router()
        window.onNavigate(ROUTES_PATH.NewBill)
      })

      afterEach(() => {
        jest.clearAllMocks()
      })
  
      test("Then if it is not an image it should display an error message and disable submit button", () => {
        const formInstance = new NewBill({ document, onNavigate, store: mockStore, localStorage: window.localStorage })
        
        const fileInput = screen.getAllByTestId("file")[0]
        const file = new File([""], "test.txt", { type: "text/plain"})
        
        const event = {
          preventDefault: jest.fn(),
          target: { files: [file], value: "C:\\fakepath\\test.txt" }
        }
        
        fileInput.dispatchEvent(new Event("change"))
        formInstance.handleChangeFile(event)
        
        expect(screen.getAllByTestId("file-error")[0].textContent).toBe("Uploadez une image en PNG, JPG ou JPEG")
        const submitButton = screen.getByTestId("submit-button")
        expect(submitButton).toBeDisabled()
      })

      test("Then if it is an image, it should enable the submit button", () => {
        const formInstance = new NewBill({ document, onNavigate, store: mockStore, localStorage: window.localStorage })
        
        const fileInput = screen.getAllByTestId("file")[0]
        const file = new File([""], "test.jpg", { type: "image/jpeg"})
        
        const event = {
          preventDefault: jest.fn(),
          target: { files: [file], value: "C:\\fakepath\\test.jpg" }
        }
        
        fileInput.dispatchEvent(new Event("change"))
        formInstance.handleChangeFile(event)
        
        const submitButton = screen.getByTestId("submit-button")
        expect(submitButton).toBeEnabled()
      })
    })

    describe("When I try to submit a new bill", () => {
      beforeEach(() => {
        Object.defineProperty(window, 'localStorage', { value: localStorageMock })

        window.localStorage.setItem('user', JSON.stringify({
          type: 'Employee'
        }))
        const root = document.createElement("div")
        root.setAttribute("id", "root")
        document.body.append(root)
        router()
        window.onNavigate(ROUTES_PATH.NewBill)
      })

      afterEach(() => {
        jest.clearAllMocks()
      })

      test("Then it should create a bill object", () => {
        let formInstance = new NewBill({ document, onNavigate, store: mockStore, localStorage: window.localStorage })

        const fileInput = screen.getAllByTestId("file")[0]
        const file = new File([""], "test.jpg", { type: "image/jpeg"})
        
        const event = {
          preventDefault: jest.fn(),
          target: { files: [file], value: "C:\\fakepath\\test.jpg" }
        }

        formInstance = {
          ...formInstance,
          fileName: "test.jpg",
          updateBill: jest.fn((bill) => bill)
        }

        console.log("filename:",formInstance.fileName);
        
        fileInput.dispatchEvent(new Event("change"))
        formInstance.handleChangeFile(event)

        const form = screen.getAllByTestId("form-new-bill")[0]
        const submitEvent = { preventDefault: jest.fn(), target: form, }
        formInstance.fileName = "test.jpg"
        formInstance.handleSubmit(submitEvent)

        const bill = {
          "id": "47qAXb6fIm2zOKkLzMro",
          "vat": "80",
          "fileUrl": "https://firebasestorage.googleapis.com/v0/b/billable-677b6.a…f-1.jpg?alt=media&token=c1640e12-a24b-4b11-ae52-529112e9602a",
          "status": "pending",
          "type": "Hôtel et logement",
          "commentary": "séminaire billed",
          "name": "encore",
          "fileName": "preview-facture-free-201801-pdf-1.jpg",
          "date": "2004-04-04",
          "amount": 400,
          "commentAdmin": "ok",
          "email": "a@a",
          "pct": 20    
        }
  
        expect(formInstance.updateBill(bill)).toBeCalled()
        // expect(submitEvent.preventDefault).toHaveBeenCalled()
      })
    })
  })


})
