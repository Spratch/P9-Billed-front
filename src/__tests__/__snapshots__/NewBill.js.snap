// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given I am connected as an employee When I am on NewBill Page When I try to submit a new bill Then it should create a bill object 1`] = `undefined`;
